csharp/src/Ice/OutgoingRequestFrame.cs

```diff
@@ -100,23 +100,70 @@ public sealed class OutgoingRequestFrame : OutputStream
                                     IReadOnlyDictionary<string, string>? context, ArraySegment<byte> payload)
             : this(proxy, operation, idempotent, context)
         {
+            var ostr = new OutputStream(Encoding, Data, PayloadStart);
             if (payload.Count == 0)
             {
-                WriteEmptyEncapsulation(_payloadEncoding);
+                ostr.WriteEmptyEncapsulation(Encoding);
+                ostr.Finish();
             }
             else
             {
-                WritePayload(payload);
+                ostr.WritePayload(payload);
+            }
```

Comment:
Shouldn't this be WriteEncapsulation on the OutputStream? I find the call to Finish here unexpected... why do we call it for one case but not the other? See my later comment about having Finish returning Tail.

Reply:
The call to Finish adjust the last segment to the written bytes, maybe `WriteEmptyEncapsulation` should do this, it is not required when we use `WritePayload` as the payload segment is already adjusted to the correct size. For Ice1 the end of the encapsulation marks the end of the stream so is fine to automatically finish the  stream. For Ice2 we will include the context after it so we don't want to adjust the last segment until we write the context.